### DISCUSSION TITLE
Barrel level monitoring using audio analysis - possible emlearn application?

### DISCUSSION THREAD

**Discussion by miketeachman** [emlearn/emlearn-micropython#27] [Ideas]
Hi!

I would like to get an opinion on the feasibility of using audio analysis to detect the level of saw dust in a metal barrel.

**Background**:  Our [Makerspace](https://www.makerspace.ca) has a wood shop with a metal dust collection barrel.  There is a [cyclone dust collector system](https://en.wikipedia.org/wiki/Cyclonic_separation) that collects sawdust created by the wood shop machines and deposits the sawdust into the barrel.  One of the challenges in our wood shop is alerting users to empty the barrel when it is about 1/2 full.  Often, no one checks the barrel and it overflows. 

**Idea for consideration**:  Use emlearn, processing audio samples from an I2S microphone (attached to the barrel) to detect the sawdust level. I am thinking that the barrel will exhibit different acoustic characteristics compared to being empty versus being full of sawdust.  Perhaps a difference in acoustic characteristics can be exploited to detect the saw dust level in the barrel? For example, maybe a "hollow sound" can be detected when the barrel is empty and a "full sound" can be detected when it is full.  The sound created by the cyclone motor should provide a driving sound into the barrel.  

Is there any merit in this idea?

Here is a photo of the metal barrel:
![drum](https://github.com/user-attachments/assets/553bf38a-a3ab-433e-a549-5c18cec7cf5e)

**Comment by jonnor** [comment]
Hi @miketeachman. Thank you so much for all the excellent work on the audio and I2S support in MicroPython. It is really super work! :)

This is quite interesting. I also a member at [a makerspace](https://github.com) with a CNC, so this is maybe relevant there also. Using audio to detect the condition of machinery is generally quite possible - and is used when other more conventional approaches like temperature and vibration is not sufficient. I actually work on such tech applied to HVAC/buildings at https://soundsensing.no :)

To evaluate the feasibility properly, one would need to collect a little bit of data. Just having one set of audio files with the following scenarios would make it possible to do a quick qualitative evaluation. It can be recorded with just a smartphone - but automatic audio gain should be turned off (since one of the most basic indicators would be change in soundlevel)!
 
- Zero fill (just emptied)
- Half full (desired trigger point)
- Full (catastrophic) . 
- Vacum not running

Presumably the CNC also makes quite a bit of sound as well. Which might get confused for extra noisy bag. So ideally, each of these would both have data for CNC on/cutting and CNC off/idle.

Actually creating a solution would need a bit more data. At minimum 10 samples of each scenario, ideally more like 100. So one could put up a device that does data collection automatically - the same kind one would use "in production" (to make the data fully representative). Since the vacuum is probably on for several minutes at a time, doing random or periodic sampling with 1-5 minute interval would probably be sufficient. Or one could connect it to the same power, to avoid capturing when it is not running.

**Comment by miketeachman** [comment]
Thanks for reviewing this idea and giving me an idea on how to proceed.  Much appreciated!  I'll start by following your suggestion to capture audio files using my iphone, for various barrel levels, and with the dust collector running.  Analyzing that sample data should give an idea if there is any signature that can be detected to derive the sawdust level in the barrel.
